File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-1.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        n = len(s)
        l, r = 0, 0
        m = {}
        res = 0

        while r < n:

            m[s[r]] = m.get(s[r], 0) + 1

            max_freq = 0
            for val in m.values():
                max_freq = max(max_freq, val)

            while (r - l + 1) - max_freq > k:
                m[s[l]] -= 1
                if m[s[l]] == 0:
                    m.pop(s[l])
                l += 1


            res = max(res, r - l + 1)
            r += 1

        return res

File: Data Structures & Algorithms/daily-temperatures/submission-1.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        stack<pair<int, int>> st;
        vector<int> res(n, 0);

        for(int i=0; i<n; i++){
            while(!st.empty() && temperatures[i] > st.top().first){
                res[st.top().second] = i - st.top().second;
                st.pop();
            }
            st.push(make_pair(temperatures[i], i));
        }

        return res;
    }
};
